### PR TITLE
Api calls implementation

### PR DESCRIPTION
**Pull request by yaswanthmyneni**
# Api calls implementation for private routes

### CODE REVIEW

File: src/components/navigation/Navigation.jsx [hunk 1]
```diff
@@ -18,6 +21,36 @@ const Navigation = () => {
...
+    const encodedToken = localStorage.getItem("token");
+    const cartList = JSON.parse(localStorage.getItem("cartList"));
+    const wishlist = JSON.parse(localStorage.getItem("wishlist"));
+    if (logout) {
+      cartList?.map((item) => {
+        return (async () =>
+          await axios({
+            method: "DELETE",
+            url: `/api/user/cart/${item._id}`,
+            headers: { authorization: encodedToken },
```
Comment (govindappaarun) [inline src/components/navigation/Navigation.jsx]: May I know the reason for calling the API in a loop?

File: src/context/cart-context.js [hunk 2]
```diff
@@ -9,6 +11,35 @@ const CartContextProvider = ({ children }) => {
...
+      cartListLocalStorage.length >= 1
+    ) {
+      cartListLocalStorage.map((item) => {
+        return (async () => {
+          const response = await axios({
+            method: "post",
+            url: `/api/user/cart`,
+            headers: { authorization: encodedToken },
+            data: {
+              product: item,
```
Comment (govindappaarun) [inline src/context/cart-context.js]: suggestion: I see this chunk of code is repeated many places, you can add some utility function for these lines as well so you can eliminate code duplication

File: src/context/wishlist-context.js [hunk 3]
```diff
@@ -1,11 +1,36 @@
...
+      wishlist.length < 1 &&
+      wishlistLocalStorage !== null &&
+      wishlistLocalStorage.length >= 1
+    ) {
+      wishlistLocalStorage.map((item) => {
+        return (async () => {
+          const response = await axios({
+            method: "post",
+            url: `/api/user/wishlist`,
+            headers: { authorization: encodedToken },
```
Comment (govindappaarun) [inline src/context/wishlist-context.js]: same comment here as well
one more suggestion: don't call the APIs in a loop

File: src/utility/cart-functions.js [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+        payload: response.data.cart,
+      });
+    })();
+  }
+};
+
+const removeFromCart = (dispatch, item) => {
+  const encodedToken = localStorage.getItem("token");
+
+  (async () => {
```
Comment (govindappaarun) [inline src/utility/cart-functions.js]: no need of async  and an iife function in #30,  is there any reason ?
instead, you can directly make removeFromCart as async

File: src/utility/cart-functions.js [hunk 5]
```diff
@@ -0,0 +1,97 @@
+import axios from "axios";
+
+const addToCart = (cartList, dispatch, item) => {
+  const encodedToken = localStorage.getItem("token");
+
+  if (cartList.find((element) => element.id === item.id)) {
+    console.log("already present");
+  } else {
+    (async () => {
```
Comment (govindappaarun) [inline src/utility/cart-functions.js]: same comment here, no need of async

File: src/utility/wishlist-functions.js [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+const addToWishlist = (wishlist, setWishlist, item) => {
+  const encodedToken = localStorage.getItem("token");
+
+  if (wishlist.find((element) => element.id === item.id)) {
+    console.log("already present");
+  } else {
+    (async () => {
+      const response = await axios({
+        method: "post",
+        url: "/api/user/wishlist",
```
Comment (govindappaarun) [inline src/utility/wishlist-functions.js]: make a utility function out of this, as its duplicated many places

**Review by govindappaarun — COMMENTED** [review]
I see some good improvements for your code
 - looks like you are not using reducer functions, 
 - understand DRY principle
 - you should not call API in a loop / map, avoid

**Review by Naman-Saxena1 — CHANGES_REQUESTED** [review]
Apart from the changes already mentioned by Arun, I was not able to find any new issues. 👍

File: src/components/navigation/Navigation.jsx [hunk 7]
```diff
@@ -18,6 +21,36 @@ const Navigation = () => {
...
+        return (async () =>
+          await axios({
+            method: "DELETE",
+            url: `/api/user/wishlist/${item._id}`,
+            headers: { authorization: encodedToken },
+          }))();
+      });
+      localStorage.clear();
+      setWishlist([]);
+      cartDispatch({ type: "cartList", payload: [] });
```
Comment (Ankur9669) [inline src/components/navigation/Navigation.jsx]: Why are you setting cartList and wishlist in localstorage as well as context

**Review by Ankur9669 — CHANGES_REQUESTED** [review]
I Agree with arun apart from that, i have also given some suggestion, which you can look to improve

**Pull request merged**